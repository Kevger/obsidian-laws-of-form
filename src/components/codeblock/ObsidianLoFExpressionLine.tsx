import {
  splitExpressionBySeparator,
  ObsidianLoFLinePartType,
} from "../../utils";
import { ObsidianLoFUninterpretedText } from "./ObsidianLoFUninterpretedText";
import { ObsidianLoFBlockErrorMessage } from "./ObsidianLoFBlockErrorMessage";
import { ObsidianLoFExpression } from "./ObsidianLoFExpression";
import "./styles.css";

/**
 * Each line of a LoF block is rendered by this component. The line is split into parts by the separator. Each part is either a text or a LoF expression. The parts are then rendered by the appropriate component.
 *
 * @param children The content of the line e.g. `()() :: = :: ()`
 * @param separator The separator used to split the line into parts e.g. `::`
 * @returns A JSX element representing the line in form of the LoF notation
 */
export function ObsidianLoFExpressionLine({
  children,
  separator,
}: {
  children?: string;
  separator: string;
}): JSX.Element {
  if (typeof children !== "string") {
    throw new Error("[LoFExpressionLine]: children must be a string");
  }

  const parts = splitExpressionBySeparator(children.toString(), separator);

  return (
    <div className="StyleObsidianLoFExpressionLine">
      {parts.length === 0 ? (
        <br key="br" />
      ) : (
        parts.map((part, i) => {
          switch (part.type) {
            case ObsidianLoFLinePartType.text:
              return (
                <ObsidianLoFUninterpretedText
                  key={`text ${part.value}${i}`}
                  text={part.value}
                  placeRight={parts.length - 1 === i}
                />
              );
            case ObsidianLoFLinePartType.lof:
              return (
                <ObsidianLoFExpression
                  key={`expr ${part.value}${i}`}
                  expression={part.value}
                />
              );
            default:
              return (
                <ObsidianLoFBlockErrorMessage
                  key={`err ${part.value}${i}`}
                  msg={"Unknown part type"}
                />
              );
          }
        })
      )}
    </div>
  );
}
