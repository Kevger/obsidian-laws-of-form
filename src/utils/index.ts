import { type MarkdownPostProcessorContext } from 'obsidian';
import { type CSSProperties } from 'react';

export interface LoFParameters {
  style: CSSProperties;
  separator: string;
}

export const enum ObsidianLoFLinePartType {
  text = 'text',
  lof = 'lof',
}

// separator to mark uninterpreted text parts
const LOF_DEFAULT_SEPARATOR = '::';

/**
 * Splits a string by a separator into an array of (type, value) objects.
 *
 * The type is either `text` or `lof` depending if the part is enclosed by the separators.
 *
 * @param text String to split, e.g. "(()) :: = :: ."
 * @param separator Separator to split the string, e.g. "::"
 * @returns Array of objects containing the type and the value
 *
 * @example
 * splitExpressionBySeparator("(()) :: = :: .", "::")
 * // returns
 * [{type: "lof", value: "(())"}, {type: "text", value: " = "}, {type: "lof", value: "."}]
 */
export function splitExpressionBySeparator(text: string, separator = '::'): Array<{ type: ObsidianLoFLinePartType, value: string; }> {
  return text
    .split(separator)
    .map((part, i) => ({ type: i % 2 === 1 ? ObsidianLoFLinePartType.text : ObsidianLoFLinePartType.lof, value: part }))
    .filter((part) => part.value.length > 0);
}

/**
 * Convert a css key string to a react key string
 *
 * kebab-case -> camelCase
 *
 * e.g. "font-size" -> "fontSize"
 *
 * @param key css key
 * @returns kebab-cased react key
 */
export function cssToReactKey(key: string): string {
  return key
    .trim()
    .split('-')
    .map((part, i) => {
      if (i === 0) return part;
      return part[0].toUpperCase() + part.slice(1);
    })
    .join('');
}

/**
 * Init the default parameters for the lof code block
 *
 * @returns an object containing the default parameters
 */
export function initLoFParameters(): LoFParameters {
  return {
    style: {},
    separator: LOF_DEFAULT_SEPARATOR
  };
}

/**
 * This function extract the optional parameters from the lof code block.
 * These parameters are defined after the lof keyword and before the first newline.
 *
 * @param ctx - the context of the code block
 * @param el - the htmlElement of the code block
 * @returns - an object containing the optional parameters
 *
 * @example
 * ````
 * ```lof font-size: 1.5em; color: red;
 * ...
 * ```
 * ````
 * returns ` { "font-size": "1.5em", "color": "red" } `
 */
export function getLoFBlockParameters(
  ctx: MarkdownPostProcessorContext,
  el: HTMLElement,
  defaultValues: LoFParameters
): LoFParameters {
  const sectionInfo = ctx.getSectionInfo(el);
  if (sectionInfo == null) {
    console.error(
      '[LoFPlugin|getLoFBlockParameters]: sectionInfo is undefined',
      sectionInfo
    );
    throw new Error('[LoFPlugin|getLoFBlockParameters]: sectionInfo is undefined');
  }

  let blockHeader = sectionInfo.text.split('\n')[sectionInfo.lineStart];

  // we need to find `lof`since a code block can start with a different number of apostrophes
  const lofKeyWordStart = blockHeader.indexOf('lof');

  if (lofKeyWordStart === -1) {
    console.error(
      '[LoFPlugin|getLoFBlockParameters]: lof keyword not found',
      blockHeader
    );
    throw new Error('[LoFPlugin|getLoFBlockParameters]: lof keyword not found');
  }

  blockHeader = blockHeader.substring(lofKeyWordStart + 3).trim();

  // use the default values first and overwrite them with possible extracted parameters
  const extractedParameters = cssStringToLoFParameters(blockHeader);
  const copyDefaultValues = deepCopy(defaultValues); // deep copy for safety
  return {
    ...copyDefaultValues,
    ...extractedParameters,
    style: {
      ...copyDefaultValues.style,
      ...extractedParameters.style
    }
  };
}

/**
 * Convert a css string e.g. "font-size: 1.5em; color: red;" to an LoFParameters object
 *
 * @param cssString the css string
 * @returns an object containing the parameters
 *
 * @example
 * cssStringToLoFParameters("font-size: 1.5em; color: red; separator: ::")
 * // returns
 * {
 *  separator: "::",
 *  style: { "fontSize": "1.5em", "color": "red" }
 * }
 */
export function cssStringToLoFParameters(cssString: string): LoFParameters {
  const parameters = cssString
    .split(';')
    .filter((parameter) => parameter.trim().length > 0);

  const parametersObject: LoFParameters = { style: {} } as any;
  parameters.forEach((parameter) => {
    const colon = parameter.indexOf(":"); // split(":") doesn't work if "::" is the separator 
    const [key, value] = [parameter.slice(0, colon), parameter.slice(colon + 1)];
    if (key.length > 0 && value.length > 0) {
      if (key.trim() === 'separator') {
        parametersObject.separator = value.trim();
        return;
      }

      (parametersObject.style as any)[cssToReactKey(key)] = value.trim();
    }
  });

  return parametersObject;
}

export function deepCopy<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj));
}
