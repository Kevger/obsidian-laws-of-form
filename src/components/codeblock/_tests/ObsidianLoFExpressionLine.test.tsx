import { type ReactTestRenderer, create as render } from "react-test-renderer";
import { ObsidianLoFExpressionLine } from "../ObsidianLoFExpressionLine";
import * as lofUtils from "../../../utils";

describe("ObsidianLoFExpressionLine", () => {
  beforeEach(() => {
    let randomMockValue = 0.000001;
    let dateMockValue = 1;
    jest.spyOn(global.Math, "random").mockImplementation(() => {
      randomMockValue = randomMockValue + 0.000001;
      return randomMockValue;
    });

    jest.spyOn(global.Date.prototype, "valueOf").mockImplementation(() => {
      return dateMockValue++;
    });
  });

  describe("given we have an expression line", () => {
    const expressionLine = "((hello world) (()))()::((abc))::()()";
    let component: ReactTestRenderer;
    beforeEach(() => {
      component = render(
        <ObsidianLoFExpressionLine separator="::">
          {expressionLine}
        </ObsidianLoFExpressionLine>
      );
    });

    it("renders correctly", () => {
      const snapshot = component.toJSON();
      expect(snapshot).toMatchSnapshot();
    });

    it("contains the the uninterpreted raw test", () => {
      expect(component.root.findByType("span").children[0]).toBe("((abc))");
    });
  });

  describe("and given we don't give a string as a child", () => {
    it("should throw an error", () => {
      expect(() =>
        render(
          <ObsidianLoFExpressionLine separator="::">
            {2 as any}
          </ObsidianLoFExpressionLine>
        )
      ).toThrow();
    });
  });

  describe("and given we have an empty string", () => {
    it("renders correctly", () => {
      const component = render(
        <ObsidianLoFExpressionLine separator="::">
          {""}
        </ObsidianLoFExpressionLine>
      );
      const snapshot = component.toJSON();
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe("and given splitExpressionBySeparator returns invalid types", () => {
    beforeEach(() => {
      jest
        .spyOn(lofUtils, "splitExpressionBySeparator")
        .mockImplementation(() => {
          return [{ type: "some invalid type" as any, value: "lorem ipsum" }];
        });
    });

    it("should display an error", () => {
      const rendered = render(
        <ObsidianLoFExpressionLine separator="::">
          {"lorem ipsum"}
        </ObsidianLoFExpressionLine>
      );
      expect((rendered.toTree() as any).rendered.rendered[0].type.name).toBe(
        "ObsidianLoFBlockErrorMessage"
      );
    });
  });
});
