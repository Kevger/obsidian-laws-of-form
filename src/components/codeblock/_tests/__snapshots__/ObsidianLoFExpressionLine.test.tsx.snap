// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObsidianLoFExpressionLine and given we have an empty string renders correctly 1`] = `
<div
  className="StyleObsidianLoFExpressionLine"
>
  <br />
</div>
`;

exports[`ObsidianLoFExpressionLine given we have an expression line renders correctly 1`] = `
<div
  className="StyleObsidianLoFExpressionLine"
>
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_4"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_1"
    >
      hello world
    </div>
     
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_3"
    >
      <div
        className="cross "
        id="Cross_1_0003cxkbtc_2"
      />
       
    </div>
     
  </div>
   
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_5"
  />
   
   
  <span
    className="StyleObsidianLoFUninterpretedText"
    style={
      {
        "marginLeft": undefined,
      }
    }
  >
    ((abc))
  </span>
  <div
    className="cross "
    id="Cross_2_00051echq_1"
  />
   
  <div
    className="cross "
    id="Cross_2_00051echq_2"
  />
   
   
</div>
`;
